### PR TITLE
Badging via gregor

### PR DESCRIPTION
**Pull request by cjb**
@keybase/react-hackers 

This PR uses Gregor's new `keybase.1.NotifyBadges.badgeState` protocol to control badging of chat messages and TLFs in the app.  It also introduces telling the service about messages we've read.  Specifically:

* when we're told of how many `newChatConversations` there are we badge the app with that number
* when we select a conversation, we mark as read up to the latest message we've seen in that conversation
* when we get an incoming message, iff it's for the currently selected conversation, we immediately mark it read

The last point above contains a race condition -- gregor's racing to tell us to badge a new chat message as we tell gregor that we've read it.  We'll need to make a (protocol) design change to handle that.  Probably instead of looking at `newChatConversations` as a global number, we badge by adding up the number of new messages in each conversation, and ignore messages from a currently selected conversation when calculating the badging number.  So we'll no longer be just using numbers from gregor.

### CODE REVIEW

**Comment by cjb** [comment]
@keybase/react-hackers bump

File: shared/actions/chat.js [hunk 1]
```diff
@@ -238,6 +236,20 @@ function * _incomingMessage (action: IncomingMessage): SagaGenerator<any, any> {
         const conversationIDKey = conversationIDToKey(incomingMessage.convID)
         const message = _unboxedToMessage(messageUnboxed, 0, yourName, conversationIDKey)
 
+        // Is this message for the currently selected conversation? If so, mark
+        // it as read ASAP to avoid badging it -- we don't need to badge, the
+        // user's looking at it already.
+        const selectedSelector = (state: TypedState) => state.chat.get('selectedConversation')
+        const selectedConversationIDKey = yield select(selectedSelector)
+        if (message && message.messageID && conversationIDKey === selectedConversationIDKey) {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: slightly more complicated than this. we only mark it as read if the app has focus. @gabriel added this to the store last week

Comment (cjb) [inline shared/actions/chat.js]: @chrisnojima @gabriel Then in this scenario:

* app doesn't have focus

* we get a message

* you focus the app to read the message

How does the message get marked as read?  (Same question for Gabriel's red line last-message-read.)

Comment (cjb) [inline shared/actions/chat.js]: Ah, thanks.  We talked about this a bit in frontend.  I don't think it's good enough for badging to wait until the next conversation select to unbadge the selected and focused messages.  So it'll be more involved to fix, by hooking up to app-level changedFocus events.

Comment (gabriel) [inline shared/actions/chat.js]: See actions/window.js if you need to do more on changedFocus...

Also you can change desktop/renderer/index.js ... `currentWindow.on('focus', () => {` to call `store.dispatch(updateBadging(` after `changedFocus`.. where updateBadging is some new action

**Comment by cjb** [comment]
@chrisnojima I've added a new commit that sagaizes actions/window and uses its ChangedFocus action to kick off an unbadging of the currently selection conversation when the window gains focus, please take another look.

@gabriel @cecileboucheron I think we should consider updating the "red line" showing the last seen message when the conversation gains focus, too.  What do you think?

**Comment by cjb** [comment]
@cecileboucheron says in private that she would like to update the orange line immediately, so I've made two more changes:

* update the orange line on focus
* require the chat tab to be the current route in routeTree before updating Chat on gaining focus.

@chrisnojima ready for another look

**Comment by cjb** [comment]
@chrisnojima Moved the unbadging from the window actions to chat actions! PTAL.

File: shared/actions/chat.js [hunk 2]
```diff
@@ -561,6 +576,31 @@ function * _selectConversation (action: SelectConversation): SagaGenerator<any,
...
+  // Reselect the current Chat conversation, to give badging a chance to
+  // update based on the focus change.
+  const selectedSelector = (state: TypedState) => state.chat.get('selectedConversation')
+  const conversationIDKey = yield select(selectedSelector)
+  const routeSelector = (state: TypedState) => state.routeTree.get('routeState').get('selected')
+  const selectedTab = yield select(routeSelector)
+  const chatTabSelected = (selectedTab === chatTab)
+
+  if (conversationIDKey && appFocused && chatTabSelected) {
+    yield put(selectConversation(conversationIDKey, true))
```
Comment (chrisnojima) [inline shared/actions/chat.js]: can we just do the badging logic and just pull that out so selecting a conversation or handling a focus changed does this. i'm worried about the side effects of selecting the conversation again (like loading more messages etc)

**Comment by cjb** [comment]
@chrisnojima Have split out updateBadging and updateLatestMessage into their own actions and reducers now.

**Pull request merged**